### DISCUSSION TITLE
Can't click hyperlink to open in web browser in eldoc-doc-buffer

### DISCUSSION THREAD

**Discussion by 488Ques** [joaotavora/eglot#1238] [Q&A]
I'm using the Gopls language server. I have this Golang code

``` golang
err = os.MkdirAll(config.FilesDir, 0755)
```

Where I place the point on os.MkdirAll method and M-x eldoc-doc-buffer to see the full documentation

![image](https://github.com/joaotavora/eglot/assets/61875656/73ba83c9-7c9b-49e6-bb27-aeef3cf8aa7c)

Then I click on the hyperlink `os.MkdirAll on pkg.go.dev` that links to the documentation website of that method, I expect the link to be opened in my web browser but instead only see the message `user-error: Nothing to follow at point` which is produced by `markdown-follow-thing-at-point`.

I'm not sure if this is intended behavior, a bug or an error on my end.

I made a temporary solution:
``` elisp
(defun eglot-open-link ()
  "Open markdown link at point in the `eldoc-doc-buffer'."
  (interactive)
  (let ((url (get-text-property (point) 'help-echo)))
    (if url
	(browse-url-xdg-open url)
      (message "No URL found at point"))))
```

**Comment by nafiz1001** [comment]
I'm getting a similar problem with rust-analyzer, except I get `progn: Beginning of buffer` as an error which is produced by the `forward-char` call in `markdown-backward-to-link-start` which is in `markdown-link-p` which is also in `markdown-follow-thing-at-point`.

I've also attempted on a temporary solution, but it doesn't work :(

**Comment by vkz** [comment]
Experiencing the exact same thing:
- eglot v1.16 from latest master via straight
- GNU Emacs 29.2 (build 1, x86_64-pc-linux-gnu) of 2024-04-02 - built from source tags/v29.2

Technically RET is bound to <mouse-2> when on the pkg.go.dev link and that is bound to `markdown-follow-thing-at-point` as per hidden properties but it either reports `Nothing to follow` or `markdown-link-p: Beginning of buffer` depending on exactly where in that link the point is.

FWIW this is in terminal emacs launched on the remote machine over ssh. ~I half expect it'll work fine in GUI Emacs on local machine somehow but I haven't tried.~

OP trick with extracting hidden url `(get-text-property (point) 'help-echo)` does work.

**Comment by ewollesen** [comment]
Same issue here:
* eglot 1.17
* GNU Emacs 29.3.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo
 version 1.18.0) of 2024-05-28

I wonder if https://github.com/joaotavora/eglot/issues/865 is somehow related? But the fix for that (PR #866) seems to have been included in eglot 1.17. There's also some mention of an issue on the eldoc side of things, so I'll look a bit at that, too.

**Comment by ParetoOptimalDev** [comment]
The issue seems to be that [markdown-backward-to-link-start](https://github.com/jrblevin/markdown-mode/blob/6d1f08b3c52ccf9129ffa069fd0bfccc66cf72d5/markdown-mode.el#L8063) uses `(skip-chars-backward "^[")`.

From an eldoc buffer where `|` denotes cursor position like:

```
encode :: forall a. ToJSON a => a -> ByteString

Defined in ‘Data.Aeson’ (aeson-2.1.2.1)

Efficiently serialize a JSON value as a lazy  L.ByteString . 

This is implemented in terms of the  ToJSON  class's  toEncoding  method.

Docum|entation <---- link

Source  <---- link
```

`(skip-chars-backward "^[")` from that position goes to the begining of the buffer and makes the next forward-char call return `forward-char: Beginning of buffer`.

So the links in the eldoc buffer aren't formatted as markdown-mode expects?

**Comment by IceAsteroid** [comment]
This problem still persists in eglot versioned 1.17.

Clicking on the link with mouse or Hitting return in keyboard on the link doesn't work to open the link, but instead, it results in marking the link.

And the highlighted text is indeed a link, I can see it by hovering the cursor above it, it shows up the external address.

![eglot-eldoc-issue](https://github.com/user-attachments/assets/a3433d52-d81e-4d56-ab0f-a6e9079bb9ba)

Tested with `emacs -q`.

**Comment by IceAsteroid** [comment]
I've come up with a workaround here in combination with @488Ques, it works

Note that I only implemented of opening links by the `RET` key, mouse clicking will not work.

Here's the code:

In my init file.
```
(defun my-setup/bind-quick-keys-in-buffers ()
  (defvar my/quick-keys-buffers-regexps
    '("\\*eglot doc\\*"
      "\\*Ement members: .*\\*"
      "\\*eldoc\\*")
    "A list of regexps of matching buffers to insert the q key as `delete-window’.
Used in `my/quick-quit-key-buffers-in-list-update-hook-setup’.")

  (defun my/quick-keys-buffers-in-list-update-hook-setup ()
    (when my/quick-keys-buffers-regexps
      ;; Concatenate regexps for buffers into one regexp in OR logic
      (let ((regexps (mapconcat
                      'identity
                      my/quick-keys-buffers-regexps
                      "\\|")))
        ;; (dolist (buffer-name (buffer-list))
        ;;   (when (buffer-match-p
        ;;          regexps
        ;;          buffer-name)
        ;;     (with-current-buffer buffer-name
        ;;       (keymap-local-set "q" 'delete-window)))
        (dolist (buf-name (buffer-list))
          (cond
           ((buffer-match-p "\\*eldoc\\*" buf-name)
            (with-current-buffer buf-name
              (keymap-local-set "RET" 'eglot-open-link)))
           ((buffer-match-p regexps buf-name)
            (with-current-buffer buf-name
              (keymap-local-set "q" 'delete-window)))
           ))))
    )
  (add-hook 'buffer-list-update-hook 'my/quick-keys-buffers-in-list-update-hook-setup)
  )
(my-setup/bind-quick-keys-in-buffers)
```

Tweaked for sole use for eldoc-buffer.
```
(defun my-setup/open-link-in-eldoc-buffer ()
  (defun my/open-link-eldoc-in-list-update-hook-setup ()
    (dolist (buf-name (buffer-list))
      (when (buffer-match-p
             "\\*eldoc*\\*"
             buf-name)
        (with-current-buffer buf-name
	         (keymap-local-set "RET" 'eglot-open-link)))
      ))
  (add-hook 'buffer-list-update-hook 'my/open-link-eldoc-in-list-update-hook-setup)
  )
(my-setup/open-link-in-eldoc-buffer)
```

**Reply by chrisbarrett** [reply]
You could simplify the above a bit by using advice; `eldoc-display-in-buffer` seems to be called to set up the eldoc buffer. This avoids the need to run code in a hook every time the buffer list is changed.

```emacs-lisp
(defun eglot-open-link ()
  (interactive)
  (if-let* ((url (get-text-property (point) 'help-echo)))
      (browse-url url)
    (user-error "No URL at point")))

(define-advice eldoc-display-in-buffer (:after (&rest _) update-keymap)
  (with-current-buffer eldoc--doc-buffer
    (keymap-local-set "RET" #'eglot-open-link)

    ;; Uncomment below if you use evil.
    ;; (general-define-key :keymaps 'local :states 'motion "RET" #'eglot-open-link)
    ))
```

**Comment by unhammer** [comment]
Here's some advice that makes both eldoc-doc-buffer and eldoc-box links clickable:

```emacs-lisp
(defun my-markdown-follow-help-or-link-at-point-advice (orig-fun &rest args)
  "Prefer to use the help-echo property as `browse-url' target."
  (let* ((event-win (posn-window (event-start last-input-event)))
         (help-echo (with-selected-frame (window-frame event-win)
                      (with-current-buffer (window-buffer event-win)
                        (get-text-property (point) 'help-echo))))
         (help-is-url (url-type (url-generic-parse-url help-echo))))
    (message "if %s (browse-url %S)" help-is-url help-echo)
    (if help-is-url
        (browse-url help-echo)
      (apply orig-fun args))))

(advice-add 'markdown-follow-link-at-point :around #'my-markdown-follow-help-or-link-advice)
```

**Comment by the42** [comment]
I cant find a proper analysis of the problem which makes it impossible to upstream:

* Is this something eldoc can be configured?
* Is this amn issue because eldoc fundamentally doesn't understand Markdown?

I tried to search the issue on emacs-devel but couldn't find anything. A vague "make it work" will not find any takers there.

EDIT: Apparently information has been sought for at least on [help-gnu-emacs](https://lists.nongnu.org/archive/html/help-gnu-emacs/2025-01/msg00290.html) but without any reply.

EDIT2: And for gopls there is this [documentation](https://github.com/golang/tools/blob/master/gopls/doc/features/passive.md#document-link)

> Emacs + eglot: not currently used.

Which reads as "not implemented" for me

**Reply by joaotavora** [reply]
Eldoc has no notion of markdown indeed. It just receives (propertized) strings, let's call them "doc nuggets". It gets these from many sources, occasionally in batches. After organizing them it passes them on to display functions, which may show them in buffers, echo area, floating frames, read them aloud, etc. Display functions can be user-provided or provided by the minor/major mode although Eglot doesn't do that. Anyway some of these nuggets in a batch may or may not be the result of markdown rendering.

One option would be to refactor the system and to tip eldoc that a particular documentation nugget in a given batch is unrendered markdown, and some Eldoc display functions (maybe some user provided one, maybe the default ones) would know about markdown (and other formats) and render it consistently, links included. There is already a mechanism in place for sources and display functions to exchange meta info about documentation nuggets. Read the docs for eldoc-display-functions.

Another option would be for the sources that do the markdown rendering currently to additionally annotate the links in the string with some extra eldoc-specific text property that specifically tells Eldoc that that text is a hyperlink. In this model, Eldoc would not know about markup formats, but some of  its default display functions would know about this specific property.

**Reply by unhammer** [reply]
In my case, the docs are getting the url in that help-echo property from running gfm-view-mode. The final buffer-string in eldoc-doc-buffer is just `Documentation`, not `[Documentation](file:///)`, but with the url in the help-echo property. However, the arg to eglot--format-markup has `[Documentation](file:///)`. Then eglot runs

`        (ignore-errors (delay-mode-hooks (funcall mode)))
` (mode being gfm-view-mode in my case) and then eglot runs

```

        (let ((inhibit-read-only t))
          (while (setq match (text-property-search-forward 'invisible))
            (delete-region (prop-match-beginning match)
                           (prop-match-end match))))
```
which deletes the (hidden) url.

Then on clicking markdown-mode tries to find the url:

```
(defun markdown-backward-to-link-start ()
  "Backward link start position if current position is in link title."
  ;; Issue #305
  (when (eq (get-text-property (point) 'face) 'markdown-link-face)
    (skip-chars-backward "^[")
    (forward-char -1)))

(defun markdown-link-p ()
  "Return non-nil when `point' is at a non-wiki link.
See `markdown-wiki-link-p' for more information."
  (save-excursion
    (let ((case-fold-search nil))
      (when (and (not (markdown-wiki-link-p)) (not (markdown-code-block-at-point-p)))
        (markdown-backward-to-link-start)
        (or (thing-at-point-looking-at markdown-regex-link-inline)
            (thing-at-point-looking-at markdown-regex-link-reference)
            (thing-at-point-looking-at markdown-regex-uri)
            (thing-at-point-looking-at markdown-regex-angle-uri))))))

```
but fails since it's deleted (there's just that remnant help-echo-property).

If I remove that deletion-part from eglot, the link in eldoc-doc-buffer is clickable without my advice (if I first select the buffer; there's still the issue that markdown seems to not pick the right window if it's not already selected, but that should be fixable in markdown-mode)..

But I guess there may be a good reason eglot deletes invisible stuff.
